### ISSUE TITLE
TurboQuant + sliding-window eviction produces incoherent output on Gemma 4

### ISSUE THREAD

**Issue by ekryski** [ekryski/mlx-swift-lm#185]
## Summary

The `TurboQuantizedKVCache` rotating-window code path (`rotatingMaxSize` / `rotatingIdx`) is implemented and works on most sliding-window-attention models (verified on Mistral 3, Ministral 3, Gemma 3) — but on **Gemma 4** it produces incoherent output. The factory dispatch in `makeAttentionCache(parameters:maxSize:keep:)` is currently held back from constructing windowed turbo caches for any model, pending investigation of the Gemma 4 interaction.

## Reproduction

On a branch where `makeAttentionCache(parameters:maxSize:keep:)` dispatches `.turbo(...)` + `maxSize` → `TurboQuantizedKVCache(... maxSize:)`:

```
$ ./scripts/benchmark.sh --model gemma4-e2b --quant 4bit --kv turbo4v2 --method simple

[TURBO] Encode kernel: WHT butterfly, dim=256, bits=2
[BENCH] Output: Except's. **SERVApa\middleware narrowed\immediately?**
```

Same script on other windowed-attention models, same code path, same TurboQuant codec:

```
$ ./scripts/benchmark.sh --model ministral3-3b --quant 4bit --kv turbo4v2 --method simple
[BENCH] Output: Hello! I'm an AI assistant—no name, just here to help with questions, tasks, clarity, or anything under the sun. Think of me as your quick, no-stress 

$ ./scripts/benchmark.sh --model gemma3-1b --quant 4bit --kv turbo4v2 --method simple
[BENCH] Output: Hello there! I'm Gemma, and I'm here to help you with various tasks! How can I assist you today?
```

## Suspected cause

Gemma 4's `newCache(parameters:)` at [`Libraries/MLXVLM/Models/Gemma4.swift:1069`](https://github.com/ekryski/mlx-swift-lm/blob/alpha/Libraries/MLXVLM/Models/Gemma4.swift#L1069) (mirror in the LLM-side) constructs a **mixed cache** when both `compressionAlgorithm: .turbo(...)` and the per-layer-type `maxSize` are set:

- **Sliding-window layers** (`layerType == "sliding_attention"`): get `maxSize = slidingWindow` (typically 4096).
- **Full-attention layers** (`layerType == "full_attention"`): get `maxSize = parameters?.maxKVSize` (typically `nil` unless the user passed `--maxKVSize`).
- **KV-shared layers** (last `numKVSharedLayers` layers): no own cache — they reuse an earlier layer's cache (in the un-skipped prefix).

Under the proposed `makeAttentionCache` factory dispatch:
- Sliding-window layers → `TurboQuantizedKVCache(maxSize: 4096)` (compressed)
- Full-attention layers → `StandardKVCache()` (uncompressed, because no maxSize)
- KV-shared layers → reuse a `TurboQuantizedKVCache` (compressed) or `StandardKVCache` (uncompressed) depending on which earlier layer they map to

Hypotheses for the corruption:
1. **Donor / shared-cache mismatch.** A KV-shared layer reuses a donor's cache; `TurboQuantizedKVCache.peek()` on a donor with `isDonor = true` is supposed to return raw fp16 K/V, but the rotating-buffer math for `peek()` may not stay in sync with `compressedWriteOffset` after rotation.
2. **Mid-prefill rotation under TurboQuant's two-phase architecture.** `TurboQuantizedKVCache` does Phase 1 (raw prefill) → Phase 2 (compressed decode) transition. If a sliding-window-bounded prefill exceeds `maxSize` *during* Phase 1, the raw → compressed transition has to honour the rotation, and `compressedWriteOffset` / `rotatingIdx` sync code at lines 1106–1121 + 1346–1364 may have edge cases that aren't reached by Mistral 3 / Ministral 3 (which use slightly different windowing parameters).
3. **Mixed compressed / uncompressed layer attention.** Gemma 4's per-layer attention math (the alpha-branch fused norm+rope path, the conditional shared-MLP gate, the KV-shared-MLP integration) might assume layer-uniform cache types. Less likely, but the audit during the WS-C consolidation flagged Gemma 4's path as the most heavily customised in the LLM tree.
4. **Boundary-skip not applied here.** `turboBoundarySkipSet(...)` is consumed by Qwen 3.5 / NemotronH but not by Gemma 4. If Gemma 4's first few attention layers are PPL-sensitive (the spec-006 doc claims this is generally the case), routing them through TurboQuant's compressed store might be the regression source independent of the rotating-window code.

## Files involved

- `Libraries/MLXLMCommon/TurboQuantKVCache.swift` — the rotating-buffer code lives at lines 750-755 (state), 884-906 (mask construction), 1106-1121 + 1219-1248 (raw → compressed transition + write paths), 1346-1364 (post-step sync). Multiple synchronisation points; subtle.
- `Libraries/MLXLMCommon/KVCacheTypes.swift:296` — `makeAttentionCache(parameters:maxSize:keep:)` is the factory currently held back.
- `Libraries/MLXVLM/Models/Gemma4.swift:1069` + `Libraries/MLXLLM/Models/Gemma4.swift` (similar) — Gemma 4's `newCache` mixes per-layer-type `maxSize`.

## Suggested investigation steps

1. **Repro on Gemma 4 LLM-side** (currently only verified on the VLM-side `gemma4-e2b`). If LLM-side also breaks, rules out VLM-specific issues.
2. **Disable Gemma 4's KV-shared layers** (set `numKVSharedLayers = 0` for the test). If that fixes it, the KV-sharing donor/peek path is suspect.
3. **Force Gemma 4 full-attention layers to also be windowed** (artificially set `parameters.maxKVSize` to match `slidingWindow`). If that fixes it, the mixed-window-and-unbounded mode is the culprit.
4. **Compare logits step-by-step** between a working model (Mistral 3 windowed turbo) and broken (Gemma 4 windowed turbo) at the first decode step, to localise which layer's K/V state diverges first.

## Recommended scope

Until investigated, the factory dispatch in `makeAttentionCache(parameters:maxSize:keep:)` should fall through to `StandardKVCache(maxSize:)` for the windowed-turbo case. The standalone `makeKVCache(scheme: .turbo, eviction: .window(size:))` factory does construct a windowed turbo and is unblocked — direct callers (tests, custom cache strategies) get it; model dispatch through the high-level factory stays safe.
